lualib/resty/events/worker.lua

```diff
@@ -104,12 +109,106 @@ function _M.new(opts)
     return setmetatable(self, _MT)
 end
 
-function _M:communicate(premature)
-    if premature then
-        -- worker wants to exit
-        return
-    end
+local function read_thread(self, broker_connection)
+    while not terminating(self) do
+        local data, err = broker_connection:recv_frame()
+        if err then
+            if not is_timeout(err) then
+                return nil, "failed to read event: " .. err
+            end
+
+            -- timeout
+            goto continue
+        end
+
+        if not data then
+            if not exiting() then
+                log(ERR, "did not receive event from broker")
+            end
+            goto continue
+        end
+
+        local event_data, err = decode(data)
+        if err then
+            if not exiting() then
+                log(ERR, "failed to decode event data: ", err)
+            end
+            goto continue
+        end
+
+        -- got an event data, push to queue, callback in events_thread
+        local _, err = self._sub_queue:push(event_data)
+        if err then
+            if not exiting() then
+                log(ERR, "failed to store event: ", err, ". data is: ",
+                         cjson_encode(event_data))
+            end
+            goto continue
+        end
+
+        ::continue::
+    end -- while not terminating
+
+    return true
+end
+
+local function write_thread(self, broker_connection)
+    local counter = 0
+
+    while not terminating(self) do
+        local payload, err = self._pub_queue:pop()
+        if err then
+            if not is_timeout(err) then
+                return nil, "semaphore wait error: " .. err
+            end
+
+            -- timeout
+            goto continue
+        end
+
+        local _, err = broker_connection:send_frame(payload)
+        if err then
```

Comment:
Yes, but that we want to log as that is really and error of sending something. E.g. we tried to send something, but that never worked. Whereas with reading we just try to read, but if there is nothing to read the read will timeout. Read needs to timeout so that we can shutdown the Kong when Kong is shutdowning. But when read timeouts it is most probably that there is just nothing to read. But when write timeouts, we definitely were writing something. or trying at least.

Reply:
Just wonder if we should print a log and continue if nothing is written (timeout).

Reply:
A write timeout feels something that indicates, that it cannot be recovered. It also feels something that should never happen.